### ISSUE TITLE
 Error reading <resource> cannot parse "${var.prepare_resource_group}" as an integer

### ISSUE THREAD

**Issue by ArseniiPetrovich** [hashicorp/terraform#18135]
### Terraform Version 0.11.7

I've tried to use `terraform plan --out terraform.tfplan` and then `terraform apply "terraform.tfplan"` and got `cannot parse "${var.prepare_resource_group}" as an integer` error. However it works if not to specify plan.

### Terraform Configuration Files
terraform.tfvars contains:
```
network_name           = "network"
environment_name       = "env"
node_count         = 1
ssh_public_key         = "../id_rsa.pub"
prepare_resource_group = 0
region                 = "East US"
```

### Expected Behavior
Infrastructure creates.

### Actual Behavior
Error `cannot parse "${var.prepare_resource_group}" as an integer` appearing

### Steps to Reproduce
1. `terraform plan --out terraform.tfplan`
2. terraform apply "terraform.tfplan"

### Additional Context
Tried to start with ansible `terraform` module and manually. None of this methods works.

### References
- #17368
- #17462

**Comment by apparentlymart** [comment]
Hi @ArseniiPetrovich! Sorry this isn't behaving properly.

The symptoms here look like a known bug that's already fixed in our development branch for the next major release, but I'm not totally sure. If you could share your configuration files (`.tf` files, rather than `.tfvars`), that will help us to reproduce the problem and verify whether it's the bug already fixed or a new issue.

Thanks!

**Comment by mengesb** [comment]
@apparentlymart Here's an example that I use:

```
* local.yp_count: local.yp_count: Error reading aws_instance.yp_eu1a count: cannot parse "${var.host_count[\"yp\"]}" as an integer
```

```hcl
variable "host_count" {
  type        = "map"
  description = "Host count"

  default = {
    yp = 1 # @$0.0960/hr LOD
  }
}

resource "aws_instance" "yp_us1a" {
  provider               = "aws.us1a"
  count                  = "${contains(var.datacenters, "us1a") ? var.host_count["yp"] : 0}"
  ami                    = "${data.aws_ami.us1a_ami.id}"
  instance_type          = "t2.nano"
  key_name               = "ssshhh"
  subnet_id              = "subnet-123456"
  vpc_security_group_ids = ["${data.aws_security_group.us1a_sg.id}"]
}

resource "aws_instance" "yp_eu1a" {
  provider               = "aws.eu1a"
  count                  = "${contains(var.datacenters, "eu1a") ? var.host_count["yp"] : 0}"
  ami                    = "${data.aws_ami.eu1a_ami.id}"
  instance_type          = "t2.nano"
  key_name               = "ssshhh"
  subnet_id              = "subnet-123456"
  vpc_security_group_ids = ["${data.aws_security_group.eu1a_sg.id}"]
}

locals {
  yp_count = "${aws_instance.yp_us1a.count + aws_instance.yp_eu1a.count}"
  yp_hosts = "${concat(aws_instance.yp_us1a.*.tags.Name, aws_instance.yp_eu1a.*.tags.Name)}"
  yp_ips   = "${concat(aws_instance.yp_us1a.*.private_ip, aws_instance.yp_eu1a.*.private_ip, aws_instance.yp_us1a.*.public_ip, aws_instance.yp_eu1a.*.public_ip)}"
}
```

**Comment by mengesb** [comment]
Here's the terraform console usage errors:
```
> "${local.yp_count}"
unknown values referenced, can't compute value
> aws_instance.yp_us1a.count
Error reading aws_instance.yp_us1a count: cannot parse "${var.host_count[\"yp\"]}" as an integer
> var.host_count["yp"]
2
```

**Comment by mengesb** [comment]
When I destroy and then run apply or plan, I get normal behavior. It's only after resources are created and I attempt to do `plan` OR `apply` do I run into that error message.

**Comment by mengesb** [comment]
Actually; I get it any time after the first apply now.

**Comment by apparentlymart** [comment]
Thanks for sharing that extra info, @mengesb!

The handling of that special `.count` pseudo-attribute in v0.11 is a little strange and, as a result, unreliable. In v0.12 (to make room for other features) the handling of this is going to change slightly: if `aws_instance.yp_us1a` has `count` set then it will appear in expressions as a list, and so `aws_instance.yp_us1a.count` will no longer be valid (lists don't have attributes).

Instead, this'll be written as `length(aws_instance.yp_us1a)`, just as you'd do for any other list. The planned configuration upgrade tool will rewrite this automatically to try to minimize the upgrade work required here.

As a result of these changes, the new approach ought to be more reliable because it's built out of the same mechanisms as everything else, rather than handled as a special case.

Unfortunately now seeing exactly what bug this is referring to I don't think we'll be able to make any sort of interim fix in 0.11 since getting this working properly for 0.12 has required some non-trivial reorganization of how Terraform handles `count`, which would be too risky to try to do in a minor release.

**Comment by apparentlymart** [comment]
Hi again @mengesb! Sorry for the long silence here.

We've now got the work I mentioned merged into master, so I modified the configuration you gave slightly so I could try it out in the v0.12.0-alpha1 prerelease build:

```hcl
provider "aws" {
  region = "us-east-1"
}

variable "host_count" {
  type        = "map"
  description = "Host count"

  default = {
    yp = 1 # @$0.0960/hr LOD
  }
}

variable "datacenters" {
  type    = list(string)
  default = ["us1a", "eu1a"]
}

resource "aws_instance" "yp_us1a" {
  count                  = "${contains(var.datacenters, "us1a") ? var.host_count["yp"] : 0}"
  ami                    = "ami-abc123"
  instance_type          = "t2.nano"
  key_name               = "ssshhh"
  subnet_id              = "subnet-123456"
  vpc_security_group_ids = ["sg-1234"]
}

resource "aws_instance" "yp_eu1a" {
  count                  = "${contains(var.datacenters, "eu1a") ? var.host_count["yp"] : 0}"
  ami                    = "ami-abc123"
  instance_type          = "t2.nano"
  key_name               = "ssshhh"
  subnet_id              = "subnet-123456"
  vpc_security_group_ids = ["sg-1234"]
}

locals {
  yp_count = "${aws_instance.yp_us1a.count + aws_instance.yp_eu1a.count}"
  yp_hosts = "${concat(aws_instance.yp_us1a.*.tags.Name, aws_instance.yp_eu1a.*.tags.Name)}"
  yp_ips   = "${concat(aws_instance.yp_us1a.*.private_ip, aws_instance.yp_eu1a.*.private_ip, aws_instance.yp_us1a.*.public_ip, aws_instance.yp_eu1a.*.public_ip)}"
}
```

As expected (for the reason I mentioned in my earlier comment), this first returned an error related to how the methodology for counting resource has changed:

```

Error: Unsupported attribute

  on count-parse-error.tf line 38, in locals:
  38:   yp_count = "${aws_instance.yp_us1a.count + aws_instance.yp_eu1a.count}"
    |----------------
    | aws_instance.yp_us1a is tuple with 1 element

This value does not have any attributes.


Error: Unsupported attribute

  on count-parse-error.tf line 38, in locals:
  38:   yp_count = "${aws_instance.yp_us1a.count + aws_instance.yp_eu1a.count}"
    |----------------
    | aws_instance.yp_eu1a is tuple with 1 element

This value does not have any attributes.
```

(that's repeated since it's reporting separately for each of the `.count` accesses in that expression; when rendered in my terminal, the offending attribute access is underlined.) So I rewrote these to be `length(aws_instance.yp_us1a)` and `length(aws_instance.yp_eu1a)` respectively, as the configuration upgrade tool will do in the final release.

That revealed a new problem:

```
Error: Attempt to get attribute from null value

  on count-parse-error.tf line 39, in locals:
  39:   yp_hosts = "${concat(aws_instance.yp_us1a.*.tags.Name, aws_instance.yp_eu1a.*.tags.Name)}"

This value is null, so it does not have any attributes.


Error: Attempt to get attribute from null value

  on count-parse-error.tf line 39, in locals:
  39:   yp_hosts = "${concat(aws_instance.yp_us1a.*.tags.Name, aws_instance.yp_eu1a.*.tags.Name)}"

This value is null, so it does not have any attributes.
```

Although the way this is reporting is a little strange (should perhaps be saying that there's no `Name` in this map), this does seem to be a valid error situation -- this would've failed in v0.11 too. To address it, I added a Name tag to each of the resource blocks.

After this, I got a successful plan:

```
Terraform will perform the following actions:

  # aws_instance.yp_eu1a[0] will be created
  + resource "aws_instance" "yp_eu1a" {
      + ami                          = "ami-abc123"
      + arn                          = (known after apply)
      + associate_public_ip_address  = (known after apply)
      + availability_zone            = (known after apply)
      + cpu_core_count               = (known after apply)
      + cpu_threads_per_core         = (known after apply)
      + get_password_data            = false
      + id                           = (known after apply)
      + instance_state               = (known after apply)
      + instance_type                = "t2.nano"
      + ipv6_address_count           = (known after apply)
      + ipv6_addresses               = (known after apply)
      + key_name                     = "ssshhh"
      + network_interface_id         = (known after apply)
      + password_data                = (known after apply)
      + placement_group              = (known after apply)
      + primary_network_interface_id = (known after apply)
      + private_dns                  = (known after apply)
      + private_ip                   = (known after apply)
      + public_dns                   = (known after apply)
      + public_ip                    = (known after apply)
      + security_groups              = (known after apply)
      + source_dest_check            = true
      + subnet_id                    = "subnet-123456"
      + tags                         = {
          + "Name" = "yp_eu1a_0"
        }
      + tenancy                      = (known after apply)
      + volume_tags                  = (known after apply)
      + vpc_security_group_ids       = [
          + "sg-1234",
        ]

      + timeouts {
        }
    }

  # aws_instance.yp_us1a[0] will be created
  + resource "aws_instance" "yp_us1a" {
      + ami                          = "ami-abc123"
      + arn                          = (known after apply)
      + associate_public_ip_address  = (known after apply)
      + availability_zone            = (known after apply)
      + cpu_core_count               = (known after apply)
      + cpu_threads_per_core         = (known after apply)
      + get_password_data            = false
      + id                           = (known after apply)
      + instance_state               = (known after apply)
      + instance_type                = "t2.nano"
      + ipv6_address_count           = (known after apply)
      + ipv6_addresses               = (known after apply)
      + key_name                     = "ssshhh"
      + network_interface_id         = (known after apply)
      + password_data                = (known after apply)
      + placement_group              = (known after apply)
      + primary_network_interface_id = (known after apply)
      + private_dns                  = (known after apply)
      + private_ip                   = (known after apply)
      + public_dns                   = (known after apply)
      + public_ip                    = (known after apply)
      + security_groups              = (known after apply)
      + source_dest_check            = true
      + subnet_id                    = "subnet-123456"
      + tags                         = {
          + "Name" = "yp_us1a_0"
        }
      + tenancy                      = (known after apply)
      + volume_tags                  = (known after apply)
      + vpc_security_group_ids       = [
          + "sg-1234",
        ]

      + timeouts {
        }
    }

Plan: 2 to add, 0 to change, 0 to destroy.
```

So with this said, it seems like the root cause of this issue is fixed in master, so I'm going to close this. It's unfortunately that this particular construct requires a change in usage for the new release, but that's important to enable other features which require `aws_instance.yp_us1a` to behave like a regular list of objects, and so we'll address that one with the configuration upgrade tool.

Thanks for reporting this weird issue, and sorry for the delay in addressing it.
